[AN-Issue-1394] Added AptosVMViewer wrapper on AptosVM for continuous view function execution

aregng opened the pull request:
- AptosVMViewer provides means to continuously execute view functions on the same state-view without recreating VM over and over again compared to AptosVM. It provides better about 7-8 times better performance compared to AptosVM. This will help to retrieve automation task information in more optimal way.

- Added Rust variant of AutomationTaskMetaData and provided means to build AutomatedTransactions based on AutomationTaskMetaData

- Enabled aptos-types unit-tests in CI flow. Ignored for the time being the tests which are currently failing.

- Added tests for newly introduced APIs/functionalities

Commit:
[AN-Issue-1394] Added AptosVMViewer wrapper on AptosVM for continous view function execution

- AptosVMViewer provides means to continously execute view functions on
the same state-view without recreating VM over and over again compared
to AptosVM. It provides better about 7-8 times better preformance
compared to AptosVM. This will help to retrieve automation
task inforamtion in more optimal way.

- Added Rust variant of AutomationTaskMetaData and provided means to
build AutomatedTransactions based on AutomationTaskMetaData

- Enabled aptos-types unit-tests in CI flow. Ignored for the time being
the tests which are currently failing.

- Added tests for newly introduced APIs/fnctionalities

sjoshisupra commented on types/src/transaction/automated_transaction.rs:
```diff
@@ -131,10 +138,213 @@ impl AutomatedTransaction {
+/// Builder interface for [AutomatedTransaction]
+#[derive(Clone, Debug, Default)]
+pub struct AutomatedTransactionBuilder {
+    /// Gas unit price threshold. Default to 0.
```
Should the default be same as chain min?

aregng replied:
Unfortunately `GAS_UNIT_PRICE` constant is not accessible in `types` crate. But taking into account that builder instance will be created generally based on the `AutomationTaskMetaData` which is from chain-state and has passed validation having gas unit price threshold to be greater than chain minimum. I assume we will be on the safe side, unless there is a major bug.

sjoshisupra replied:
@aregng , ok for now we can leave it. We should develop `chain_info.move` module that exposes chain parameters to Move layers (read-only). But that could be a separate issue/project.

sjoshisupra commented on types/src/transaction/automated_transaction.rs:
```diff
@@ -131,10 +138,213 @@ impl AutomatedTransaction {
+    /// stored in the sender's account at the time the transaction executes.
+    pub(crate) sequence_number: Option<u64>,
+
+    /// The transaction payload, e.g., a script to execute.
```
Did you mean _an entry function to execute_ ? since we are not going to support script payload for automation

sjoshisupra reviewed (commented):
Minor comment/question

Commit:
Fixed a comment

Commit:
Fixed transaction parent hash calculation to match txn hash in smr-moonshot

- smr moonshot uses Keccak::256 to calcualte the hash of the data.
  Introduced `HashValue::keccak_256_of` API and utilized it to calcuate
  `TransactionMetadata::txn_app_hash`

aregng committed:
[AN-Issue-1432] Convenience updates to facilitate supra-layer on automation tasks execution (#154)

Co-authored-by: Aregnaz Harutyunyan <>

aregng committed:
Merge pull request #149 from Entropy-Foundation/fix/txn-app-hash

Fixed transaction parent hash calculation to match txn hash in smr-moonshot
